ipaserver/plugins/cert.py

```diff
@@ -1643,21 +1643,26 @@ def _ca_search(self, raw, pkey_only, exactly, **options):
         result = collections.OrderedDict()
         complete = bool(ra_options)
 
-        # workaround for RHBZ#1669012
-        # Improve performance for service and host case by also searching
-        # for subject. This limits the amount of certificate retrieved from
-        # Dogtag. The special case is only used, when no ra_options are set
-        # and exactly one service or host is supplied.
-        # The complete flag is left to False.
+        # workaround for RHBZ#1669012 and RHBZ#1695685
+        # Improve performance for service, host and user case by also
+        # searching for subject. This limits the amount of certificate
+        # retrieved from Dogtag. The special case is only used, when
+        # no ra_options are set and exactly one service, host, or user is
+        # supplied.
+        # IPA enforces that subject CN is either a hostname or a username.
+        # The complete flag is left to False to catch overrides.
```

Comment:
This code (including the existing workaround) makes some assumptions that I'm not 100% sure about - in particular that the hostname/username is used as the CN in the issued cert (a custom profile may violate this assumption).  I need to work through it in more detail, and have run out of week, so I'll resume early next week.

Reply:
I'm confident that the restriction 
https://github.com/freeipa/freeipa/blob/350954589774499d99bf87cb5631c664bb0707c4/ipaserver/plugins/cert.py#L747-L753 won't allow a custom profile to create certs that do have a different CN.

Reply:
Agreed. I had the same concerns as @frasertweedale but after looking at the code I don't see a way for an IPA-issued cert to not have cn=uid in the subject.

Reply:
@tiran @rcritten it is actually pretty easy - you just import a profile configuration with parameter like:
```
policyset.serverCertSet.1.default.params.name=UID=$request.req_subject_name.cn$, O=EXMAPLE.COM
```

The CSR comes in with `CN=alice`, and comes out with `UID=alice` (but of course how you configure the subject DN to look is arbitrary).

Reply:
@frasertweedale Oh no :/

We can either decide to ignore this edge case or somebody has to redesign and rewrite the entire ``cert_find`` API. The current implementation doesn't scale at all. Several common cases download **all** certificates from Dogtag, which puts a lot of stress on 389-DS, Dogtag, and the Python API. The BZ mentions > 100k certs.

I have an idea how to optimize the most critical cases for cert revocation. We can use the fact that host certs, service certs, and user certs are always stored on the host, service, or user LDAP entry. For simple host, service, and user cert search, perform ``_ldap_search`` before ``_ca_search``. In ``_ca_search`` only retrieve certificates with subject CNs or serial numbers of certs as returned by ``_ldap_search``.

There is one catch: It looks like Dogtag does not support list of serial numbers. :/

Reply:
We can add support on the Dogtag side for that.  I'm inclined to ACK this, given the problem only occurs on an edge case, and file tickets to rework the whole thing.

Reply:
... once you fix the `'users'` / `'user'` issue that is :)